pkg/common/log/reopen.go

```diff
@@ -0,0 +1,89 @@
+package log
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+const (
+	fileFlags = os.O_APPEND | os.O_CREATE | os.O_WRONLY
+	fileMode  = 0640
+)
+
+var _ ReopenableWriteCloser = (*ReopenableFile)(nil)
+
+type (
+	// Reopener inspired by https://github.com/client9/reopen
+	Reopener interface {
+		Reopen(*Logger) error
+	}
+	ReopenableWriteCloser interface {
+		Reopener
+		io.WriteCloser
+	}
+)
+
+type (
+	ReopenableFile struct {
+		name        string
+		f           *os.File
+		unsafeClose unsafeClose
+		mu          sync.Mutex
+	}
+	// unsafeClose must be called while holding the lock
+	unsafeClose func() error
+)
+
+func NewReopenableFile(name string) (*ReopenableFile, error) {
+	file, err := os.OpenFile(name, fileFlags, fileMode)
+	if err != nil {
+		return nil, err
+	}
+	return &ReopenableFile{
+		name:        name,
+		f:           file,
+		unsafeClose: file.Close,
```

Comment:
As written, the unsafeClose function will only ever close the originally opened file and won't result in subsequent log files being closed in later rotations. If we want to continue to hook this functionality, then this I'd suggest we change this to:

```
func closeFile(f *os.File) error {
    return f.Close()
}

type ReopenableFile struct {
    ...
    closeFunc func(*os.File) error
}

func NewReopenableFile(name string) (*ReopenableFile, error) {
    ...
    return &ReopenableFile{
        closeFunc: closeFile,
    }
}
```

Then tests can replace the closeFunc on the returned *ReopenableFile to inject the error.

Reply:
Good catch. I thought I had solved the test injection. So close (pun intended).